[Woo POS] Pagination support: Initial infinite scrolling

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of https://github.com/woocommerce/woocommerce-ios/issues/14186
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR adds initial support for infinite scrolling to the Point of Sale, for this, we observe the changes on the view's height and attempt to fetch the next page of products.

If there are new products, we append them to the existing list and continue keeping track of the current page for subsequent pagination attempts.

We also update the pull-to-refresh to empty whatever is in memory and reset the page count to zero. There are still some improvements to be done, tracked in the [main issue](https://github.com/woocommerce/woocommerce-ios/issues/14186).

https://github.com/user-attachments/assets/f89c0757-bd1f-4a4a-a0f0-40e8a8c22125

## Testing information
- For easiness, update `POSConstants.productsPerPage` to a lower number of products that would still be larger than the available screen space, for example "10":
- In POS, observe that scrolling down keeps loading new products
- Without exiting POS, in the web, create a new eligible product and publish it, then in POS pull-to-refresh. Observe that items in memory are entirely cleared, and we're back to page 0 with the newest item at the top.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
nextPage initial logic. fetch next on tap

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Automate next page load on scroll

iamgabrielma committed:
ammend ab1e3db7 make compile

iamgabrielma committed:
Clean up for review

iamgabrielma committed:
Assure unique productIDs. Test current page behavior

iamgabrielma committed:
Remove newitems count check

For the moment we do not need to track if it’s the last page, we can have a better solution when we support more product types. If we do, we also would need to handle the state = .empty case here, so merchants can see the correct error screen.

iamgabrielma committed:
clear all items in memory on reload

jaclync commented on WooCommerce/Classes/POS/ViewModels/ItemListViewModel.swift:
```diff
@@ -150,6 +161,5 @@ private extension ItemListViewModel {
             value: "Retry",
             comment: "Text for the button appearing on the item list screen when there's an error loading products."
         )
-        static let firstPageNumber: Int = 1
```
nit: since this is specific to the WC API (starting at 1 whereas it might start at 0 for other API), it might be nice having this as a constant. WDYT of keeping this constant and setting `currentPage`'s default value (in two places - variable declaration and `reload`) to `firstPageNumber - 1`?

iamgabrielma replied:
Yes, this is next on my [list](https://github.com/woocommerce/woocommerce-ios/issues/14186), since I want to make a clear separation between first fetch vs subsequent fetches. 

At the moment we call `populatePointOfSaleItems()` both when the view first appears, as well as in subsequent fetches for pagination, so right now fetching the 1st page correctly is a bit of a side-effect of the internal function count and relies that we start on 0. Ideally I think we should pass the page into the function as a parameter and keep it stateless, but was a bit of too much refactoring right now to keep this PR small.

jaclync commented on WooCommerce/Classes/POS/ViewModels/ItemListViewModel.swift:
```diff
@@ -10,6 +10,8 @@ final class ItemListViewModel: ItemListViewModelProtocol {
     @Published private(set) var isHeaderBannerDismissed: Bool = false
     @Published var showSimpleProductsModal: Bool = false
 
+    @Published private(set) var currentPage: Int = 0
```
❓ Is this variable public just for testing? I think generally we keep variables private whenever possible. Maybe the assertions can be made on the `itemProvider.providePointOfSaleItems`'s `pageNumber`?

iamgabrielma replied:
It seems it is 😅 , I'm going to leave it like this for now based on the next changes described here https://github.com/woocommerce/woocommerce-ios/pull/14195/files?diff=unified&w=0#r1813903775 as I think that's what we're going to test, otherwise I'll make it fully private.

> Maybe the assertions can be made on the itemProvider.providePointOfSaleItems's pageNumber?

I'm curious, could you clarify how could we test this bit? Do you mean for example asserting if we get the correct `simulatedResponse` for different pages?

jaclync replied:
> I'm curious, could you clarify how could we test this bit? Do you mean for example asserting if we get the correct simulatedResponse for different pages?

In `ItemListViewModelTests`, it's already using a mock item provider `MockPOSItemProvider`. We can potentially add a variable to `MockPOSItemProvider` for the latest page number, set by `providePointOfSaleItems`, and the test cases can have assertions on this page number property.

jaclync commented on WooCommerce/Classes/POS/ViewModels/ItemListViewModel.swift:
```diff
@@ -37,13 +39,20 @@ final class ItemListViewModel: ItemListViewModelProtocol {
+            let uniqueNewItems = newItems.filter { newItem in
+                !items.contains(where: { $0.productID == newItem.productID })
+            }
+            items.append(contentsOf: uniqueNewItems)
```
💬 Another thought for optimization (please ignore if this is already a future task): if the `items` count is less than the page size, it's safe to assume we've reached the last page so that no new API requests are made.

iamgabrielma replied:
That sounds reasonable, specially for the current limit of 100 items, I found this not working well when we have a large gap of unsupported items, but that would be an edge case and less prone to happen as we increase the number of supported products.I've [added it](https://github.com/woocommerce/woocommerce-ios/issues/14186) to the main issue along other optimization items 🙇

jaclync replied:
> That sounds reasonable, specially for the current limit of 100 items, I found this not working well when we have a large gap of unsupported items, but that would be an edge case and less prone to happen as we increase the number of supported products.

If there are items filtered and hidden from the items in the API response, I imagine this check on whether there's a next page would be done on the items in the API response like in `POSProductProvider.providePointOfSaleItems` and then passed to the view model level here.

jaclync commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -133,6 +133,20 @@ private extension ItemListView {
             }
             .padding(.bottom, floatingControlAreaSize.height)
             .padding(.horizontal, Constants.itemListPadding)
+            .background(GeometryReader { proxy in
+                Color.clear
+                    .onChange(of: proxy.frame(in: .global).maxY) { maxY in
+                        if viewModel.state == .loading {
+                            return
+                        }
+                        let viewHeight = UIScreen.main.bounds.height
+                        if maxY < viewHeight {
+                            Task {
+                                await viewModel.populatePointOfSaleItems()
+                            }
+                        }
+                    }
+            })
```
non-blocking: maybe this could be extracted to a modifier to detect scrolling to a threshold?

iamgabrielma replied:
It can! It's on the list [here](https://github.com/woocommerce/woocommerce-ios/issues/14186)

jaclync reviewed (approved):
Works as expected! 🚢

iamgabrielma commented:
Thanks for the review! I'll tackle the page declaration feedback in the next PR with the upcoming changes to separating first fetch from subsequent fetches, and logged the rest separately. I'll merge this for the time being!
